File: CODE/SelfJoinBuildWinLoss.R
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: SelfJoinBuildWinLoss.R
# DATE 2025-10-13
# DESCRIPTION: 
# This R program is a data enrichment pipeline for a basketball database. 
# Its primary purpose is to calculate team scores and determine game winners, 
# then save those results back into the database and a CSV file.

library(RSQLite)
library(DBI)
library(dplyr)
library(readr)
library(stringr)

# Connect to database
path_to_database <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
con <- dbConnect(RSQLite::SQLite(), path_to_database)

# Check if 'PTS' column exists in the 'Boxscores' table
if (!("PTS" %in% dbListFields(con, "Boxscores"))) {
  dbDisconnect(con)
  stop("Error: The 'PTS' variable is not present." +  
       "Please run ComputePercentagesPoints program again.", 
    call. = FALSE)
}

# Create scores by summarizing Boxscores
scores <- tbl(con, "Boxscores") |>
  group_by(GAME_ID, TEAM_ID) |>
  summarise(
    SCORE = sum(PTS),
    .groups = "drop"
  )

# Left join scores with itself
query <- scores |>
      left_join(scores, 
                by = "GAME_ID", 
                suffix = c("_team","_opponent"), 
                relationship = "many-to-many")   |>
        filter(TEAM_ID_team != TEAM_ID_opponent) |>
          mutate(
                WIN_LOSS_team = case_when(
                  SCORE_team > SCORE_opponent ~ 1,
                  SCORE_team < SCORE_opponent ~ 0
                )
                ) |>
            mutate(
                    WIN_LOSS_opponent = case_when(
                      SCORE_opponent > SCORE_team ~ 1,
                      SCORE_opponent < SCORE_team ~ 0
                    )
                  ) |>
              select(-ends_with("_opponent")) |>
                rename_with(~ str_remove(., "_team"), ends_with("_team")) 

# Display the query
show_query(query) 

# Run the query
results_df <- query |>
  collect()

boxscores_df <- tbl(con, "Boxscores") |> collect()

joined_df <- left_join(boxscores_df, results_df, by = c("GAME_ID", "TEAM_ID")) 

print(joined_df, width = Inf)

# Put the new data back into the database: overwrite the Boxscores table
dbWriteTable(
  con,
  name = "Boxscores",
  value = joined_df,
  overwrite = TRUE
)

write_csv(joined_df, "simple_boxscores.csv")

# Disconnect database connection
dbDisconnect(con)

File: CODE/SelfJoinBuildWinLoss.py

# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: SelfJoinBuildWinLoss.py
# DATE 2025-10-19
# DESCRIPTION: 
# This program updates a database by calculating game scores and 
# win/loss outcomes for basketball teams, then merging those results 
# back into the main dataset. It essentially automates the "standings" 
# calculation for a set of boxscore data.

import sqlite3
import pandas as pd
import os
import sys

# Connect to database
path_to_database = os.getenv("EXAMPLES", "") + "Boxscores.db"
con = sqlite3.connect(path_to_database)

# Create a cursor object
cursor = con.cursor()

# Check if the 'PTS' column exists in the 'Boxscores' table
table_name = "Boxscores"
column_to_check = "PTS"

try:
    # Use PRAGMA table_info to get column information
    cursor.execute(f"PRAGMA table_info({table_name})")
    columns_info = cursor.fetchall()
    
    # Extract column names from the result
    # info[1] is the column name
    column_names = [info[1] for info in columns_info] 
    
    if column_to_check not in column_names:
        print(f"Error: The '{column_to_check}' column is not present." 
              "Run ComputePercentagesPoints program again.")
        con.close()
        sys.exit(1) # Exit the program with an error code
        
except sqlite3.Error as e:
    print(f"Database error: {e}")
    con.close()
    sys.exit(1)

# Create scores by summarizing Boxscores
# This step is not actually needed since the subsequent query does 
# the same thing.
scores_df = pd.read_sql_query("""
    SELECT GAME_ID, TEAM_ID, SUM(PTS) AS SCORE
    FROM Boxscores GROUP BY GAME_ID, TEAM_ID""", con)

# Use a standard SQL query to perform the self-join and conditional logic.
# This query is already correct and creates the final results_df.
query_sql = """
    WITH scores AS (
        SELECT GAME_ID, TEAM_ID, SUM(PTS) AS SCORE
        FROM Boxscores
        GROUP BY GAME_ID, TEAM_ID
    )
    SELECT
        t1.GAME_ID,
        t1.TEAM_ID AS TEAM_ID,
        t1.SCORE AS SCORE,
        CASE
            WHEN t1.SCORE > t2.SCORE THEN 1
            WHEN t1.SCORE < t2.SCORE THEN 0
            ELSE NULL
        END AS WIN_LOSS
    FROM scores AS t1
    LEFT JOIN scores AS t2
        ON t1.GAME_ID = t2.GAME_ID
    WHERE t1.TEAM_ID != t2.TEAM_ID;
"""

# Execute the query and load results into a DataFrame
results_df = pd.read_sql_query(query_sql, con)

# Join results back to original boxscores and write to database
boxscores_df = pd.read_sql_query("SELECT * FROM Boxscores", con)

# Drop existing 'SCORE' and 'WIN_LOSS' columns if they exist in 
# boxscores_df 
columns_to_drop = ['SCORE', 'WIN_LOSS']
for col in columns_to_drop:
    if col in boxscores_df.columns:
        boxscores_df = boxscores_df.drop(columns=col)

# Perform the left join using pandas
# The 'SCORE' and 'WIN_LOSS' columns from results_df will be added to the 
# boxscores_df
joined_df = pd.merge(boxscores_df, results_df, 
                     on=['GAME_ID', 'TEAM_ID'], how='left')

# Write the new data back into the Boxscores table, overwriting it
joined_df.to_sql(name='Boxscores', con=con, if_exists='replace', index=False)

# Disconnect database connection
con.close()

print(joined_df)

joined_df.to_csv("boxscore.csv")
